### PR TITLE
Add draft post about build ID

### PR DESCRIPTION
**Pull request by franc0is**

### CODE REVIEW

File: _drafts/gnu-build-id-for-firmware.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+As you can see, the `.text`, `.bss`, `.data`, `.stack` sections each have the 
+`LOAD` attribute, all others (including our `.note.gnu.build-id`) do not and will be
+discarded.
+
+To add a section to our binary, we must specify an address for it in our linker
+script. Assuming your linker script declares the following memory layout:
+
+```
+MEMORY¬
```
Comment (chrisc11) [inline _drafts/gnu-build-id-for-firmware.md]: I'd probably drop the newline markers `¬` so someone can just copy/paste

Comment (franc0is) [reply _drafts/gnu-build-id-for-firmware.md]: Fixed

File: _drafts/gnu-build-id-for-firmware.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+Versions are a great way to identify a *release*: a set of interfaces and
+capabilities bundled together.
+
+Versions do not - however - identify a specific binary all that well. For
+example, you could have multiple binaries for a given version in order to
+accommodate different variants of your hardware in the field.
+
+For this, we need something else. This is where the GNU build ID comes in.
+
+Why would we want to identify a specific binaries? A few cases:
```
Comment (chrisc11) [inline _drafts/gnu-build-id-for-firmware.md]: "a specific binaries" -> "a specific binary"

File: _drafts/gnu-build-id-for-firmware.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    Build ID: bab6b09f86b3c3017499d8e386447a610c559bd5
+```
+
+As you can see, our binary now contains the build id
+`bab6b09f86b3c3017499d8e386447a610c559bd5`.
+
+## Bundling the GNU build ID in firmware bin files
+
+Getting the build ID in your executables on Linux is easy: they are ELF files!
+Firmware on the other hand typically deals with binaries which are assembling by
```
Comment (chrisc11) [inline _drafts/gnu-build-id-for-firmware.md]: `assembling` -> `assembled` ?

File: _drafts/gnu-build-id-for-firmware.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+This is typically accomplished with objcopy:
+```terminal
+$ arm-none-eabi-objcopy firmware.elf firmware.bin -O binary
+```
+
+This takes every elf section earmarked to be loaded and places them at 
+the correct offset in the bin file. In the process, most debug sections are stripped out.
+
+Dumping the elf sections of the resulting `minimal.elf` gives us:
```
Comment (chrisc11) [inline _drafts/gnu-build-id-for-firmware.md]: Maybe include what tool /CLI command you used to dump. This looks like `objdump -h`?

File: _drafts/gnu-build-id-for-firmware.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+ 12 .ARM.attributes 0000002c  00000000  00000000  0001707a  2**0
+                  CONTENTS, READONLY
+ 13 .debug_frame  00000298  00000000  00000000  000170a8  2**2
+                  CONTENTS, READONLY, DEBUGGING
+```
+
+As you can see, our build ID now has an address assigned to it and is marked
+with the `LOAD` attribute.
+
+**Note**: Make sure to declare the `.gnu_build_id` section after the `.text`
```
Comment (chrisc11) [inline _drafts/gnu-build-id-for-firmware.md]: You could back reference to your first blog post if you want here

File: _drafts/gnu-build-id-for-firmware.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        printf("%02x", build_id_data[i]);
+    }
+    printf("\n");
+}
+```
+
+## References
+
+- [Original build ID proposal](https://fedoraproject.org/wiki/RolandMcGrath/BuildID)
+- [Elf Note section description](https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-18048.html)
```
Comment (martijnthe) [inline _drafts/gnu-build-id-for-firmware.md]: For fun perhaps also point to the ld sources:
https://github.com/bminor/binutils-gdb/blob/4de283e4b5f21207fe12f99913d1f28d4f07843c/ld/emultempl/elf32.em#L1165

Comment (franc0is) [reply _drafts/gnu-build-id-for-firmware.md]: sweet, yes!

**Pull request merged**